external-crates/move/crates/move-command-line-common/src/error_bitset.rs

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) The Diem Core Contributors
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+const TAG_MASK: u64 = 0x8000_0000_0000_0000;
+const ERROR_CODE_MASK: u64 = 0x7fff_0000_0000_0000;
+const LINE_NUMBER_MASK: u64 = 0x0000_ffff_0000_0000;
+const IDENTIFIER_INDEX_MASK: u64 = 0x0000_0000_ffff_0000;
+const CONSTANT_INDEX_MASK: u64 = 0x0000_0000_0000_ffff;
+
+pub struct ErrorBitset {
+    // |<tagbit>|<error code>|<line number>|<identifier index>|<constant index>|
+    //   1-bit      15-bits      16-bits        16-bits          16-bits
+    pub bits: u64,
+}
+
+impl ErrorBitset {
+    pub fn new(
+        line_number: u16,
+        error_code: u16,
+        identifier_index: u16,
+        constant_index: u16,
+    ) -> Self {
+        let mut bits = 0u64;
+        bits |= 1u64 << 63;
+        // OK to shift over by 48 because we know the error code is only 15 bits and therefore will
+        // not affect the tag bit.
+        bits |= (error_code as u64) << 48;
+        bits |= (line_number as u64) << 32;
+        bits |= (identifier_index as u64) << 16;
+        bits |= constant_index as u64;
+        Self { bits }
+    }
+
+    pub fn from_u64(bits: u64) -> Option<Self> {
+        if Self::is_tagged_error(bits) {
+            Some(Self { bits })
+        } else {
+            None
+        }
+    }
+
+    pub fn is_tagged_error(bits: u64) -> bool {
+        (bits & TAG_MASK) >> 63 == 1
+    }
```

Comment:
Maybe a `get_bits` with all of these as `const`s?

```suggestion
    pub fn is_tagged_error(bits: u64) -> bool {
        get_bits(TAG_MASK, TAG_SHIFT) == 1
    }
```

No worries if not so.

Reply:
Will check it out :) I think it might be nice to actually lift both the masks and  shifts into an enum and then set things up on that.

Reply:
Updated the logic here to be a bit more general/reuse as much logic as possible. So I think it should be gtg now :)